### ISSUE TITLE
[New Project] Proposal for a kubectl plugin for Confidential Containers

### ISSUE THREAD

**Issue by surajssd** [confidential-containers/confidential-containers#140]
## Project Description

As the project is becoming feature rich, there are more and more things that a user needs to do before they start deploying applications to Kubernetes using confidential containers.

And as these new features are emerging so are the instructions to use them, hence there is a fragmentation of tooling around it. Right now, the policy proposal has a tool that clients have to use to annotate their pods/deployments before deploying. The confidential data hub proposal as it is emerging it proposes that users "ecnrypt" data and put it in a secret. Also, right now we don't have a single tool command-line-user-side tool to give a unified experience of using confidential containers.

## Alignment with Confidential Containers

As a project that is tightly coupled with Kubernetes it is fair to assume that users will be using kubectl to deploy their applications. Hence it is a good idea to have a kubectl plugin that will help users to use confidential containers' features from a single place.

Kubectl plugins are generally used as kubectl subcommands like the following:

```bash
kubectl coco ...
kubectl cc ...
```

The plugin will provide the following facilities:

- Use the existing policy generation tool and have a command to generate policies.

```bash
kubectl coco policygen ...
```

- Various approaches to create secrets / sealed secrets. This fits in perfectly with the [proposal on supporting sealed secret](https://github.com/confidential-containers/guest-components/issues/295).

```bash
kubectl coco secretgen ...
```

- A subcommand to upload secrets to KBS (in its current avatar):

```bash
kubectl coco upload-to-kbs ...
```

## Advantages

- An unified experience for user of confidential containers.
- A platform/framework for us to add more customer side features.

## Challenges

- I don't know the state of the k8s's rust SDK. This one seems pretty well-maintained: https://kube.rs/

## Implementation

To create a kubectl plugin we need to create a binary that is named as `kubectl-coco` and put in the `$PATH`. Kubectl invokes this binary with a call to `kubectl coco ...`.

## GitHub repositories needed

Either one of these, or as folks suggest more:

- kubectl-coco
- kubectl-cc

**Comment by surajssd** [comment]
cc: @mkulke @bpradipt @danmihai1

**Comment by Xynnn007** [comment]
Great. I'm wondering if image signing tool can also be one? Potentially, image encryption tool.

**Comment by arronwy** [comment]
cool, very useful feature/tool especially for our CoCo end users.

**Comment by bpradipt** [comment]
@surajssd great idea. This will also help to manage specifics about cloud-api-adaptor (peer-pod) for example viewing the pod-vm association, resource consumption details etc.
https://github.com/confidential-containers/cloud-api-adaptor/issues/188

**Comment by bpradipt** [comment]
As for the project name, my first preference will be `kubectl-coco` to align with CoCo project.

**Comment by mythi** [comment]
> cool, very useful feature/tool especially for our CoCo end users.

+1

**Comment by mkulke** [comment]
I wonder whether some aspects of a locked-down control-plane could be addressed by such a tool, like a `kubectl exec` (`kubectl coco exec?`) that doesn't undermine the tcb?

**Comment by mkulke** [comment]
> Great. I'm wondering if image signing tool can also be one? Potentially, image encryption tool.

Indeed, codifying/packaging conventions for image enc into an accessible command would be great addition.

**Comment by ray-valdez** [comment]
I think this approach can also be leveraged in the [split API work](https://github.com/confidential-containers/confidential-containers/issues/53) to create a plugin to communicate directly with the `kata-agent` running inside of the confidential VM to manage the tenant workload.

**Comment by katexochen** [comment]
> I don't know the state of the k8s's rust SDK. This one seems pretty well-maintained: https://kube.rs/

Maybe it would make sense to use Go for such a plugin instead?

**Comment by mkulke** [comment]
> > I don't know the state of the k8s's rust SDK. This one seems pretty well-maintained: https://kube.rs/
> 
> Maybe it would make sense to use Go for such a plugin instead?

down the road this might create some work if we want to use some of the project's rust libraries. fwiw there is [prior art](https://crates.io/crates/kubectl-view-allocations) for kubectl plugins in rust

**Comment by Xynnn007** [comment]
@bpradipt I am glad that you are open to adopt trustee-cli abilities to the plugin. So let me state some observations.

1. I see existing code on both sides, one in Rust and the other in Go.

2. Both tools share the goal of better using Trustee.

3. kubectl-coco includes the ability to deploy CoCo on the cluster side, while trustee-cli does not.

My suggestions are:

1. Ideally, coordinate @spotlesstofu with to avoid duplication of effort and have a better co-operation. We need an architecture design which I think engineers are always excited to do. The two codebases should ideally be grouped together. kubectl-coco should functionally have at least two submodules: one serving the Trustee side and one serving the CoCo runtime/kubernetes side. Trustee-cli can provide abilities of the Trustee-side sub-function implementation for kubectl-coco.

2. Discuss how to implement interaction: should it involve forking different binaries (i.e., not implementing them as the same binary file or using the same language), or should it be designed as a library to integrate trustee-cli into kubectl-coco?

3. How to design CI test and deliveries and interact with the Kata side to test the entire functional flow? cc @stevenhorsman @fidencio 

4. Consider user-friendly documentation and update it on the CoCo website.

cc @fitzthum @mkulke @mythi

**Comment by bpradipt** [comment]
The more I think of it, having this kind of an opinionated workflow for the app developer will be valuable 

1. Deploy Trustee server in the specific developer namespace
2. Any app deployed in the specific namespace will use the Trustee server for attestation. All K8s secrets in the manifest will be automatically converted to sealed secrets and provisioned in the Trustee server.

The only config required will kubeconfig with access to specific namespace. 
The goal here is for app developer to  CoCo-fy the app and test it.

@Xynnn007 @fitzthum @mkulke  and others, wdyt ?

**Comment by fitzthum** [comment]
Cool. Let's try to get this on the agenda for an upcoming Thursday meeting and bring this into the org.

I do wonder about finding maintainers for this, especially since it is in Go and has a fair amount of code already created, but generally I am in favor.
